### PR TITLE
Replay renderer: add line-rendering and unproject()

### PR DESCRIPTION
**Pull request by eundersander**
## Motivation and Context

Adding some features (and hacks) to our experimental replay renderers, needed for a Python application for SIRo.

* integrate DebugLineRender into replay renderers
* add ClassicReplayRenderer.unproject
* stub implementation for BatchReplayRenderer.unproject
* lighting hack in ClassicReplayRenderer to get ReplicaCAD stages to render correctly

## How Has This Been Tested

Local testing.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/esp/bindings/SimBindings.cpp [hunk 1]
```diff
@@ -454,7 +454,11 @@ void initSimBindings(py::module& m) {
           [](AbstractReplayRenderer& self) {
             return py::capsule(self.getCudaColorBufferDevicePointer());
           },
-          R"(Retrieve the depth buffer as a CUDA device pointer.)");
+          R"(Retrieve the depth buffer as a CUDA device pointer.)")
+      .def("debug_line_render", &AbstractReplayRenderer::getDebugLineRender,
+           R"(Get visualization helper for rendering lines.)")
+      .def("unproject", &AbstractReplayRenderer::unproject,
```
Comment (eundersander) [inline src/esp/bindings/SimBindings.cpp:460]: Some context: I'm writing a SIRo Python application that uses this. That app sets up a division between the renderer and the sim, paving the way for these pieces to live in separate processses or separate machines across a network (client/server architecture). Eventually, we need a way for sim code to unproject from screen space to world space, e.g. given a mouse cursor position (or arbitrary screen position), compute a world ray. I'd eventually like a replay renderer to provide an `unprojectHelper` object that encapsulates the camera matrix (including projection/viewport params). This object can be serialized and passed back to the sim, where the sim can do as many arbitrary `unproject`s as it wants. For now, as a bandaid, I'm having the renderer unproject the mouse cursor position, call this helper, and return the world ray to the sim.

See the unproject usage here (WIP):
https://github.com/eundersander/habitat-lab/blob/eundersander/magnum_glfw_interactive_play/habitat-lab/habitat/utils/gui_app_wrapper.py#L189

TLDR: this API should eventually go away in favor of a `getUnprojectHelper` API.

File: src/esp/bindings/SimBindings.cpp [hunk 2]
```diff
@@ -454,7 +454,11 @@ void initSimBindings(py::module& m) {
           [](AbstractReplayRenderer& self) {
             return py::capsule(self.getCudaColorBufferDevicePointer());
           },
-          R"(Retrieve the depth buffer as a CUDA device pointer.)");
+          R"(Retrieve the depth buffer as a CUDA device pointer.)")
+      .def("debug_line_render", &AbstractReplayRenderer::getDebugLineRender,
```
Comment (eundersander) [inline src/esp/bindings/SimBindings.cpp:458]: Context: see my earlier comment about separate sim and renderer. We eventually need a way for the sim to convey lines to the renderer. For now in my SIRo Python app, since both pieces live in the same thread, I'm simply giving the sim direct access to the `debug_line_render` object where it can directly add lines. See here:
https://github.com/eundersander/habitat-lab/blob/eundersander/magnum_glfw_interactive_play/habitat-lab/habitat/utils/gui_app_wrapper.py#L219

File: src/esp/sim/BatchReplayRenderer.cpp [hunk 3]
```diff
@@ -274,6 +278,24 @@ void BatchReplayRenderer::doRender(
...
+    debugLineRender_->flushLines(projCamMatrix, renderer_->tileSize());
+  }
+}
+
+esp::geo::Ray BatchReplayRenderer::doUnproject(
+    unsigned envIndex,
+    const Mn::Vector2i& viewportPosition) {
+  // temp stub implementation: produce a placeholder ray that varies with
+  // viewportPosition
+  return esp::geo::Ray(
```
Comment (eundersander) [inline src/esp/sim/BatchReplayRenderer.cpp:295]: Another shortcut here. We can implement this properly later. As is, this will be useful for helping @0mdc integrate the batch replay renderer into the SIRo Python app.

File: src/esp/sim/ClassicReplayRenderer.cpp [hunk 4]
```diff
@@ -31,6 +31,10 @@ ClassicReplayRenderer::ClassicReplayRenderer(
   resourceManager_ =
       std::make_unique<assets::ResourceManager>(metadataMediator, flags);
 
+  // hack to get ReplicCAD non-baked stages to render correctly
+  resourceManager_->getShaderManager().setFallback(
```
Comment (eundersander) [inline src/esp/sim/ClassicReplayRenderer.cpp:35]: Long-term, I hope @0mdc can figure out why the lighting isn't working correctly for the classic replay renderer.

Comment (mosra) [reply src/esp/sim/ClassicReplayRenderer.cpp:35]: I remember I was doing some material patching for the ReplicaCAD composites, in particular [adding or removing the flat/unlit material bit](https://github.com/facebookresearch/habitat-sim/blob/07d0b3069068446c886cccb9fae8cffaf0783f2f/src/utils/compositor/compositor-replicacad.cpp#L230-L239) (not sure which was the case exactly), because otherwise both the prebaked and non-baked arch files were rendered with the same shader, looking wrong.

So if this works for you for the batch renderer & my composites but not the classic one, that might be the reason.

File: src/esp/sim/BatchReplayRenderer.cpp [hunk 5]
```diff
@@ -256,6 +257,9 @@ void BatchReplayRenderer::doRender(
                  "with a standalone renderer", );
   static_cast<gfx_batch::RendererStandalone&>(*renderer_).draw();
 
+  // todo: integrate debugLineRender_->flushLines
+  CORRADE_INTERNAL_ASSERT(!debugLineRender_);
```
[the remark this answers is not in the document]
Comment (mosra) [reply src/esp/sim/BatchReplayRenderer.cpp:261]: On my "eventually nice to have" TODO list I have a batch-friendly line generator for the new [line shader](https://doc.magnum.graphics/magnum/classMagnum_1_1Shaders_1_1LineGL.html) (right now it's only possible to generate & draw a line mesh as a whole, not gradually add to it). 

Depending on how soon you'd like to have it working for the batch renderer, I can prioritize it.

[the remark this answers is not in the document]
Comment (eundersander) [reply src/esp/sim/BatchReplayRenderer.cpp:261]: Related FYI, I can't think of a use case where we'll need line-rendering for a multi-tile (multi-env) setup. I.e. thick lines are *only* for showing to humans, not AI model training, because they are fundamentally a UX thing and not photorealistic. They're only for single-tile usage of the batch renderer.

File: src/esp/sim/AbstractReplayRenderer.h
Comment (0mdc) [inline src/esp/sim/AbstractReplayRenderer.h:1]: It would be great to include a screenshot-based test for line rendering and unit test for unproject.
I don't mind doing this in future changesets as this code is not fully developed.

File: src/esp/sim/AbstractReplayRenderer.h [hunk 6]
```diff
@@ -108,6 +113,17 @@ class AbstractReplayRenderer {
   // Retrieve the depth buffer as a CUDA device pointer. */
   virtual const void* getCudaDepthBufferDevicePointer();
 
+  std::shared_ptr<esp::gfx::DebugLineRender> getDebugLineRender(
+      unsigned envIndex);
+
+  esp::geo::Ray unproject(unsigned envIndex,
```
Comment (0mdc) [inline src/esp/sim/AbstractReplayRenderer.h:119]: It would be helpful to comment that this is a viewport point to world-space ray function.

**Review by 0mdc — APPROVED** [review]
LGTM!

File: src/esp/sim/AbstractReplayRenderer.h
Comment (eundersander) [reply src/esp/sim/AbstractReplayRenderer.h:1]: I added a unit test for `unproject`. An image test for line-rendering will probably be super-flaky, because thick-line support is flaky in OpenGL. @mosra has indicated that "real" thick-line rendering will be coming to Magnum eventually, so we could do an image test at that time.
